### DISCUSSION TITLE
Optional keys to support model with DB defaults

### DISCUSSION THREAD

**Discussion by Dreamsorcerer** [pydantic/pydantic#2809] [Question]
We have pydantic models which validate an object before saving to the DB and also after loading them from a DB. The DB has some server defaults, so some fields are optional, but will always be present when reading from the DB.

This is currently represented with something like:
```
class Foo(BaseModel):
    id: Optional[int] = None
    bar: str
```

The problem is that this produces many mypy errors relating to `Foo.id` being `None`, even though it would never be `None` when reading from the DB.

The simplest solution to me would be to make the keys optional, but I don't see a way to do this from the docs. i.e. The equivalent of:
```
class _Foo(TypedDict, total=False):
    id: int
class Foo(_Foo):
    bar: str
```

I don't see a way to do this, the docs seem to mix up `typing.Optional` and optional keys, so there doesn't appear to be a way to define a key as optional, only keys which default to `None`. Unless I'm misunderstanding something.

**Comment by PrettyWood** [comment]
Hi @Dreamsorcerer 
You're right it's not directly possible. See https://github.com/samuelcolvin/pydantic/discussions/2353 for a suggested change for v2.
In the meantime you can just use some kind of sentinel as default value for `Unset`
